commitment_scheme/src/merkle/hash.rs

```diff
@@ -0,0 +1,84 @@
+use ark_ff::{BigInteger, PrimeField};
+use blake2::{Blake2s256, Digest};
+use felt::Felt252;
+use generic_array::typenum::U32;
+use generic_array::GenericArray;
+use poseidon::{FieldHasher, Poseidon3};
+use sha3::Keccak256;
+use std::{fmt::Debug, marker::PhantomData};
+
+pub trait Hasher<F: PrimeField> {
+    type Output: Clone + Eq + Default + Debug;
+
+    // compress a list of internal nodes into a single internal node
+    fn node(input: &[Self::Output]) -> Self::Output;
+
+    // compress a list of leaves into a single leaf
+    fn leaf(input: &[F]) -> Self::Output;
+}
+
+pub struct Blake2s256Hasher<F: PrimeField> {
+    _ph: PhantomData<F>,
+}
+
+impl<F: PrimeField> Hasher<F> for Blake2s256Hasher<F> {
+    type Output = GenericArray<u8, U32>;
+
+    fn leaf(input: &[F]) -> Self::Output {
+        let mut hasher = Blake2s256::new();
+        input
+            .iter()
+            .for_each(|f| hasher.update(f.into_bigint().to_bytes_be()));
+        hasher.finalize()
+    }
+
+    fn node(input: &[Self::Output]) -> Self::Output {
+        let mut hasher = Blake2s256::new();
+        input.iter().for_each(|f| hasher.update(f));
+        hasher.finalize()
+    }
+}
+
+pub struct Keccak256Hasher<F: PrimeField> {
+    _ph: PhantomData<F>,
+}
+
+impl<F: PrimeField> Hasher<F> for Keccak256Hasher<F> {
+    type Output = GenericArray<u8, U32>;
+
+    fn leaf(input: &[F]) -> Self::Output {
+        let mut hasher = Keccak256::new();
+        input
+            .iter()
+            .for_each(|f| hasher.update(f.into_bigint().to_bytes_be()));
+        hasher.finalize()
+    }
+
+    fn node(input: &[Self::Output]) -> Self::Output {
+        let mut hasher = Keccak256::new();
```

Comment:
Yes, the `Hasher` trait implementation is same for `Blake2s256Hasher` and `Keccak256Hasher`. I thought of making it generic for all hashes but the trait implementation of `Poseidon3Hasher` is different from the other two, so I am not sure if it can be made generic over all hashes.